tt_metal/impl/buffers/circular_buffer_types.cpp

```diff
@@ -0,0 +1,187 @@
+}
+
+CircularBufferConfig::Builder::Builder(CircularBufferConfig &parent, uint8_t buffer_index) :
+    parent_(parent), buffer_index_(buffer_index) {
+    if (buffer_index > NUM_CIRCULAR_BUFFERS - 1) {
+        TT_THROW(
+            "Buffer index ({}) exceeds max number of circular buffers per core ({})",
+            buffer_index,
+            NUM_CIRCULAR_BUFFERS);
+    }
+    parent_.buffer_indices_.insert(buffer_index_);
+}
+
+CircularBufferConfig::Builder &CircularBufferConfig::Builder::set_data_format(const tt::DataFormat &data_format) {
+    parent_.data_formats_[buffer_index_] = data_format;
+    return *this;
+}
+
+CircularBufferConfig::Builder &CircularBufferConfig::Builder::add_size(uint32_t size) {
+    parent_.total_size_ += size;
+    return *this;
+}
+
+CircularBufferConfig::Builder &CircularBufferConfig::Builder::set_page_size(uint32_t page_size) {
+    if (parent_.total_size_ % page_size != 0) {
+        TT_THROW("Total circular buffer size {} B must be divisible by page size {} B", parent_.total_size_, page_size);
+    }
+    if (page_size % sizeof(uint32_t) != 0) {
+        TT_THROW("Page size must be divisible by sizeof(uint32_t) because buffers hold uint32_t values");
+    }
+    parent_.page_sizes_[buffer_index_] = page_size;
+    return *this;
+}
+
+CircularBufferConfig::Builder &CircularBufferConfig::Builder::set_tile_dims(const Tile &tile) {
+    parent_.tiles_[buffer_index_] = tile;
+    return *this;
+}
+
+CircularBufferConfig::Builder CircularBufferConfig::index(uint8_t buffer_index) { return Builder(*this, buffer_index); }
+
+void CircularBufferConfig::set_config(const std::map<uint8_t, tt::DataFormat> &data_format_spec) {
+    if (data_format_spec.size() > NUM_CIRCULAR_BUFFERS) {
+        TT_THROW(
+            "Only {} circular buffer slots are available but data formats are specified for {} indices",
+            NUM_CIRCULAR_BUFFERS,
+            data_format_spec.size());
+    }
+
+    for (const auto &[buffer_index, data_format] : data_format_spec) {
+        if (buffer_index > NUM_CIRCULAR_BUFFERS - 1) {
+            TT_THROW(
+                "Buffer index ({}) exceeds max number of circular buffers per core ({})",
+                buffer_index,
+                NUM_CIRCULAR_BUFFERS);
+        }
+        this->data_formats_[buffer_index] = data_format;
+        this->buffer_indices_.insert(buffer_index);
+    }
+}
+
+bool operator==(const CircularBufferConfig &lhs, const CircularBufferConfig &rhs) {
+    if (lhs.total_size() != rhs.total_size() || lhs.globally_allocated_address() != rhs.globally_allocated_address() ||
+        lhs.data_formats() != rhs.data_formats() || lhs.page_sizes() != rhs.page_sizes() ||
+        lhs.tiles() != rhs.tiles()) {
+        return false;
+    }
+
+    if (lhs.shadow_global_buffer && rhs.shadow_global_buffer) {
+        return lhs.shadow_global_buffer == rhs.shadow_global_buffer;
+    }
+
+    return !lhs.shadow_global_buffer && !rhs.shadow_global_buffer;
```

Comment:
As far as I can tell the following is overly complicated
```
    if (lhs.shadow_global_buffer && rhs.shadow_global_buffer) {
        return lhs.shadow_global_buffer == rhs.shadow_global_buffer;
    }

    return !lhs.shadow_global_buffer && !rhs.shadow_global_buffer;
```

and it can be replaced with
```
return lhs.shadow_global_buffer == rhs.shadow_global_buffer;
```

If not, why not?

Reply:
done